### ISSUE TITLE
separated gql fragments on different type in a union should generate union type instead intersection type

### ISSUE THREAD

**Issue by frantic1048** [dotansimha/graphql-code-generator#695]
*Sorry for the complicated title, I've no idea how to describe this issue shorter in one sentence. （´＿｀）*

**Describe the bug**

When a named fragment is referenced outside a query, the generated type of the fragments is an intersection of fragments, which should be a union.

**To Reproduce**

*schema*:

```gql
# schema.graphql
#
# mostly copied from:
# https://facebook.github.io/graphql/June2018/#example-255de

union SearchResult = Photo | Person

type Person {
  name: String
  age: Int
}

type Photo {
  height: Int
  width: Int
}

type Query {
  firstSearchResult: SearchResult
}
```

*document*:

```gql
# document.graphql

fragment PersonFragment on Person {
  name
}

fragment PhotoFragment on Photo {
  height
}

query search {
  firstSearchResult {
    ...PersonFragment
    ...PhotoFragment
  }
}
```

*Run command*: `gql-gen --template graphql-codegen-typescript-template --schema schema.graphql document.graphql --out schema.ts`

**Expected behavior**

In generated `schema.ts`, the actual type of `query search` is:

```ts
export namespace Search {
  export type Variables = {};

  export type Query = {
    __typename?: "Query";
    firstSearchResult?: FirstSearchResult | null;
  };

  export type FirstSearchResult = PersonFragment.Fragment &
    PhotoFragment.Fragment;
}
```

Notice the `FirstSearchResult` type, it is an intersection type of `PersonFragment.Fragment` and `PhotoFragment.Fragment`.

Expected  `FirstSearchResult` type should be union of `PersonFragment.Fragment` and `PhotoFragment.Fragment`. i.e.

```ts
export type FirstSearchResult = PersonFragment.Fragment |
   PhotoFragment.Fragment;
```

**Environment:**

 - OS: Arch Linux (latest)
 - Codegen: 0.12.6
 - Node: 10.11.0
 - graphql: 14.0.2

**Additional context**

If we inline the fragments in `document.graphql`(just like the example in https://facebook.github.io/graphql/June2018/#example-f8163 ):

```gql
query search {
  firstSearchResult {
    ... on Person {
      name
    }
    ... on Photo {
      height
    }
  }
}
```

It will not run into this issue.

However, we use separated fragments mostly because:

- The fragments are too much
- The fragments are not as simple as the example shows
- We want to reuse fragments and ensure maintainability

I think this is a reasonable use case.

About the reason of this issue, I think the parser is lacking the context of a fragment reference(do not know the referenced fragment is for a specific type or not, i.e. with or without `on XXXType`). Maybe the root cause is in upstream `facebook/graphql` (need to check the output of `parse` function).

**Comment by dotansimha** [comment]
That's definitely a bug. I'll look into it soon.

**Comment by dotansimha** [comment]
fixed in 0.13.0. @frantic1048
